platform/linux-generic/drv_driver.c

```diff
@@ -532,10 +579,239 @@ odpdrv_devio_t odpdrv_devio_register(odpdrv_devio_param_t *param)
+ * -The enumerator's API (name and version) requested by D is provided
+ * by the enumerator which enumerated the device.
+ * This function will return zero if the above condition where met by some
+ * devio D and the driver probe function returns 0 (success).
+ * The function will return -1 if some devio D were found, but the driver
+ * returned a non-zero value when probed (for all of them).
+ * The function will return -2 if no devio matching the above requirement was
+ * found.
+ * The function will return -3 if the device was already bound to a driver */
+static int probe_device_driver(_odpdrv_device_t *dev, _odpdrv_driver_t *drv)
+{
+	int i;
+	int ret = -2;
+	_odpdrv_devio_t *devio;
+	_odpdrv_enumr_t *enumr;
+	_odpdrv_enumr_class_t *enumr_c;
+
+	/* the device already has a driver?: end of story... */
+	if (dev->driver)
+		return -3;
+
+	/* look at the different devio this driver can work with: */
+	for (i = 0; i < ODPDRV_MAX_DEVIOS; i++) {
+		/* look at each registered devios: */
+		devio_list_read_lock();
+		for (devio = devio_lst.head; devio; devio = devio->next) {
+			/* if devio is no good for this driver, keep searching*/
+			if ((strncmp(drv->param.devios[i].api_name,
+				     devio->param.api_name,
+				     ODPDRV_NAME_SIZE) != 0) ||
+			    (drv->param.devios[i].api_version !=
+			     devio->param.api_version))
+				continue;
+
+			/* give a chance to the devio to reject the device
+			 * if it feels it should do so: */
+			if (devio->param.probe &&
+			    devio->param.probe((odpdrv_device_t)dev))
+				continue;
+
+			/* grab the device enumerator and its class: */
+			enumr = get_enumr(dev->param.enumerator);
+			enumr_c = get_enumr_class(enumr->param.enumr_class);
+
+			/* if devio is no good for this dev, keep searching */
+			if ((strncmp(devio->param.enumr_api_name,
+				     enumr->param.api_name,
+				     ODPDRV_NAME_SIZE) != 0) ||
+				     (devio->param.enumr_api_version !=
+					enumr->param.api_version))
+				continue;
+
+			/* seems we are good to probe the driver: */
+			odp_ticketlock_lock(&drv->probelock);
+			if (drv->param.probe((odpdrv_device_t)dev,
+					     (odpdrv_devio_t)devio, i) == 0) {
+				/* the driver accepts this device */
+				odp_ticketlock_unlock(&drv->probelock);
+				devio_list_read_unlock();
+				ODP_DBG("driver %s will handle device %s(%s)\n",
+					drv->param.name,
+					dev->param.address,
+					enumr_c->param.name);
+				dev->driver = drv;
+				dev->devio = devio;
+				return 0;
+			}
+			odp_ticketlock_unlock(&drv->probelock);
+
+			/* driver did not accept the device: keep searching */
+			ret = -1;
+		}
+		devio_list_read_unlock();
+	}
+	return ret;
+}
+
+/* an empty callback is given to the driver on unprobe, if no real callback is
+ * needed */
+static void empty_unbind_callback(odpdrv_device_t odpdrv_dev ODP_UNUSED)
+{
+}
+
+/* unbind the device driver from the device (i.e. "unprobe")
+ * if the immediate flag is set, the unbind is requested to be immediate,
```

Comment:
Yes, here the flags derived from odpdrv_device_destroy() should be passed to driver->unbind(), not only the immediate flag bit only, it is like that unbind_device_driver() truncate the flags into one bit only.

If no obstacle I'll change the unbind_device_driver() to pass all flags instead of immediate bit only.